server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,382 @@
+                                } else if (zipEntry != null) {
+                                    // new entry, so write the entry header
+                                    zipOutputStream.putNextEntry(zipEntry);
+                                    zipEntry = null;
+                                } else {
+                                    // writing entry body
+                                    if (bodyPart.isPartComplete() == false) {
+                                        try (var innerChunk = bodyPart.encodeChunk(sizeHint, recycler)) {
+                                            final var iterator = innerChunk.iterator();
+                                            BytesRef bytesRef;
+                                            while ((bytesRef = iterator.next()) != null) {
+                                                zipOutputStream.write(bytesRef.bytes, bytesRef.offset, bytesRef.length);
+                                            }
+                                        }
+                                    }
+                                    if (bodyPart.isPartComplete()) {
+                                        if (bodyPart.isLastPart()) {
+                                            zipOutputStream.closeEntry();
+                                            transferReleasable(releasables);
+                                            final var newQueueLength = queueLength.decrementAndGet();
+                                            if (newQueueLength == 0) {
+                                                // next entry isn't available yet, so we stop iterating
+                                                isPartComplete = true;
+                                                assert getNextPart == null;
+                                                getNextPart = continuationListener::addListener;
+                                            } else {
+                                                // next entry is immediately available so start sending its chunks too
+                                                final var nextEntry = entryQueue.poll();
+                                                assert nextEntry != null;
+                                                zipEntry = nextEntry.zipEntry();
+                                                bodyPart = nextEntry.firstBodyPart();
+                                                releasable = nextEntry.releasable();
+                                            }
+                                        } else {
+                                            // this body part has a continuation, for which we must wait
+                                            isPartComplete = true;
+                                            assert getNextPart == null;
+                                            getNextPart = l -> bodyPart.getNextPart(l.map(p -> new QueueConsumer(null, p)));
```

Comment:
The lifecycle of `QueueConsumer` is a bit surprising to me. We reuse the same one across Entry boundary but creating a new one once across Part boundary. I wonder whether it would be easier to reason to always have a new instance when crossing either of these boundaries?

Reply:
This is deliberate, we don't want to introduce extra pauses here if none of the entries is pausing. But if one of the entries has a pause then it's ok to preserve that pause.